src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -268,7 +268,17 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             return $a == $b ? 0 : 1;
         });
 
-        if ($extra && $query = http_build_query($extra, '', '&')) {
+        // extract fragment
+        $fragment = isset($extra['_fragment']) ? $extra['_fragment'] : '';
+        unset($extra['_fragment']);
+
+        if (defined('PHP_QUERY_RFC3986')) {
+            $query = http_build_query($extra, '', '&', PHP_QUERY_RFC3986);
+        } else {
+            $query = http_build_query($extra, '', '&');
```

Comment:
Actually, I'm hesitant to deviate from using `PHP_QUERY_RFC3986`, as this is now in line with how Twig handles route generation: https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Extension/Core.php#L615